tools/ReadConsoleInputStream/ReadConsoleInputStream.cs

```diff
@@ -0,0 +1,194 @@
+    public sealed class ReadConsoleInputStream : Stream
+    {
+        private const int BufferSize = 256;
+        private const int BytesPerWChar = 2;
+        private readonly BlockingCollection<Kernel32.INPUT_RECORD> _nonKeyEvents;
+        private IntPtr _handle;
+
+        /// <summary>
+        /// Creates an instance of ReadConsoleInputStream over the standard handle for StdIn. 
+        /// </summary>
+        /// <param name="nonKeyEvents">A BlockingCollection provider/consumer collection for collecting non key events.</param>
+        public ReadConsoleInputStream(BlockingCollection<Kernel32.INPUT_RECORD> nonKeyEvents) :
+            this(Kernel32.GetStdHandle(Kernel32.StdHandleType.STD_INPUT_HANDLE), nonKeyEvents)
+        {
+        }
+
+        /// <summary>
+        /// Creates an instance of ReadConsoleInputStream over a caller-provided standard handle for stdin. 
+        /// </summary>
+        /// <param name="handle">A HFILE handle representing StdIn</param>
+        /// <param name="nonKeyEvents">A BlockingCollection provider/consumer collection for collecting non key events.</param>
+        internal ReadConsoleInputStream(HFILE handle,
+            BlockingCollection<Kernel32.INPUT_RECORD> nonKeyEvents)
+        {
+            Debug.Assert(handle.IsInvalid == false, "handle.IsInvalid == false");
+
+            _handle = handle.DangerousGetHandle();
+            _nonKeyEvents = nonKeyEvents;
+        }
+
+        public override bool CanRead { get; } = true;
+
+        public override bool CanWrite => false;
+
+        public override bool CanSeek => false;
+
+        public override long Length => throw new NotSupportedException("Seek not supported.");
+
+        public override long Position
+        {
+            get => throw new NotSupportedException("Seek not supported.");
+            set => throw new NotSupportedException("Seek not supported.");
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            _handle = IntPtr.Zero;
+
+            base.Dispose(disposing);
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            ValidateRead(buffer, offset, count);
+
+            Debug.Assert(offset >= 0, "offset >= 0");
+            Debug.Assert(count >= 0, "count >= 0");
+            Debug.Assert(buffer != null, "bytes != null");
+
+            // Don't corrupt memory when multiple threads are erroneously writing
+            // to this stream simultaneously.
+            if (buffer.Length - offset < count)
+                throw new IndexOutOfRangeException("IndexOutOfRange_IORaceCondition");
+
+            int bytesRead;
+            int ret;
+
+            if (buffer.Length == 0)
+            {
+                bytesRead = 0;
+                ret = Win32Error.ERROR_SUCCESS;
+            }
+            else
+            {
+                var charsRead = 0;
+                bytesRead = 0;
+
```

Comment:
why track both? It looks like there's a lot of effort going into keeping track of this twice in two variables that need to be kept in sync and needing that *2 multiplier constant.

perhaps just track the bytes only and you could do things like:
`buffer[offset + bytesRead++] = lsb
buffer[offset + bytesRead++] = msb` 
and drop the *2 thing all together

Reply:
It's more about giving a little more clarity in the logic (for the reader) in exchange for a little less brevity. There's a time and a place for optimization but if you think I should squash these vars, I'll do it.

Reply:
I've seen this get screwed up so many times in the `conhost.exe` code that I don't generally let anyone get away with it anymore internally, so I'm on the fence.

On the one hand, this isn't conhost. It's an example for people to understand how this works and this way is definitely clearer to read.

On the other hand, people love to copy paste, forget about it for 10 years, then someone completely different comes by and tries to patch something in the middle and desynchronizes the variables without noticing the impact and the tests of course don't catch it.

Usually when I solve this problem pattern, I abstract it into an object that has methods for incrementing or returning the size in either domain without revealing the loose math. But that's probably overkill here.

I don't want to make work for you when you're already being so generous with this sample, so I'd be satisfied with just a warning comment here to future editors about keeping these two in sync if the loop below ever evolves.

Reply:
I'll put in a disclaimer comment above the block saying that it could be optimized but it's been left more explicit for clarity. Does that work?

Reply:
Yeah that works. Thanks for bearing with my thought process.